devito/types/basic.py

```diff
@@ -938,9 +942,6 @@ def __init_finalize__(self, *args, **kwargs):
         # There may or may not be a `Grid`
         self._grid = kwargs.get('grid')
 
-        # A `Distributor` to handle domain decomposition
-        self._distributor = self.__distributor_setup__(**kwargs)
```

Comment:
why did u move it?

Reply:
Iirc, with functions on subdomains, the halo setup needs to be aware of the decomposition, so the distributor needs setting up before the halo